### PR TITLE
feat: dagster lib for retl syncs

### PR DESCRIPTION
**Pull request by RanjeetMishra**
## Description of the change

1. Dagster lib to run job for RETL syncs.
2. Resources: Currently RETLResource added, others can be added for ETL/Profiles later.
3. Basic op to trigger and wait for a sync is added so that one can define jobs. More ops like, just trigger a sync and not wait can be added, but have not seen this in many dagster apps.
4. Added tests for resource, ops and job run.

part of REV-1134
part of REV-1106

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: rudder_dagster/__init__.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from dagster import Definitions
+from rudder_dagster.resources.rudderstack import RudderStackRETLResource
+from rudder_dagster.jobs.jobs import rs_retl_job
+
+defs = Definitions(
```
Comment (RanjeetMishra) [inline rudder_dagster/__init__.py:5]: This file will be empty for the lib. Currently this inits definitions for dagster UI for local testing.

File: rudder_dagster/jobs/jobs.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+from dagster import job, RunConfig, Definitions
```
Comment (RanjeetMishra) [inline rudder_dagster/jobs/jobs.py:1]: We will not have this folder for jobs in lib. Only in Readme how to schedule jobs with configuration for resource and op.

File: README.md [hunk 3]
```diff
@@ -1,2 +1,47 @@
-# rudder-dagster
-A Dagster library for triggering reverse ETL syncs in RudderStack
+# rudder_dagster
+
+This is a [Dagster](https://dagster.io/) project scaffolded with [`dagster project scaffold`](https://docs.dagster.io/getting-started/create-new-project).
```
Comment (RanjeetMishra) [inline README.md:3]: ReadMe will be updated post PyPi lib and testing usage.

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from setuptools import find_packages, setup
+
+setup(
+    name="rudder_dagster",
+    packages=find_packages(exclude=["rudder_dagster_tests"]),
+    install_requires=[
+        "dagster",
```
Comment (vyeshwanth) [inline setup.py:7]: It's a good idea to set the minimum version for dependencies like `>=1.0.0`

File: setup.cfg [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[metadata]
```
Comment (vyeshwanth) [inline setup.cfg:1]: I am seeing 3 configuration files `pyproject.toml`, `setup.py`, `setup.cfg`

we can just use one. I suggest `pyproject.toml`. It has everything we need for this project

https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#a-full-example

File: setup.cfg [hunk 6]
```diff
@@ -0,0 +1,2 @@
...
+name = rudder_dagster
```
Comment (vyeshwanth) [inline setup.cfg:2]: also the package name, we can follow the convention like how other libraries name. I suggest `dagster-rudderstack` ? 

All the integrations use `dastster-<integration>` format for naming
https://docs.dagster.io/integrations

File: rudder_dagster/resources/rudderstack.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
+import logging
+import requests
+import time
+from abc import abstractmethod
+from dagster import ConfigurableResource, Failure, get_dagster_logger
+from dagster._utils.cached_method import cached_method
+from importlib.metadata import PackageNotFoundError, version
+from pydantic import Field
```
Comment (vyeshwanth) [inline rudder_dagster/resources/rudderstack.py:8]: we are using requests, pydantic. we should have these in our list of required dependencies.

File: rudder_dagster/resources/rudderstack.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    ) -> RudderStackRetlOutput:
+        """Triggers a sync and keeps polling till it completes.
+
+        Args:
+            conn_id (str): connetionId for an RETL sync.
+            sync_type (str): (optional) full or incremental. Default is incremental.
+        Returns:
+            RudderStackRetlOutput: Details of the sync run.
+        """
+        self._log.info(f"Tigger sync for connectionId: {conn_id} and wait for finish")
```
Comment (vyeshwanth) [inline rudder_dagster/resources/rudderstack.py:197]:
```suggestion
        self._log.info(f"Trigger sync for connectionId: {conn_id} and wait for finish")
```

typo: `trigger`

**Comment by vyeshwanth** [comment]
`rudder_dagster/resources/temp` do we need this file?

File: rudder_dagster/resources/rudderstack.py (hunk 7 above)
Comment (RanjeetMishra) [reply rudder_dagster/resources/rudderstack.py]: Added as dependencies in pyproject.toml

File: setup.cfg (hunk 5 above)
Comment (RanjeetMishra) [reply setup.cfg:1]: Based on file details, thinking to keep setup.py and setup.cfg.
https://docs.dagster.io/guides/understanding-dagster-project-files

Note: pyproject.toml was introduced in [PEP-518](https://peps.python.org/pep-0518/) and meant to replace setup.py, but we may still include a setup.py for compatibility with tools that do not use this spec.

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+[build-system]
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "dagster_rudderstack"
+version = "1.0.0"
```
Comment (RanjeetMishra) [inline pyproject.toml:7]: @vyeshwanth Not sure if this right way to manage version release. We can sync up once how we want to manage versioned release.

File: dagster_rudderstack/resources/rudderstack.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        """Polls for completion of a sync.
+
+        Args:
+            conn_id (str): connetionId for an RETL sync.
+            sync_type (str): (optional) full or incremental. Default is incremental.
+        Returns:
+            Dict[str, Any]: Parsed json output from syncs endpoint.
+        """
+        status_endpoint = f"/v2/retl-connections/{conn_id}/syncs/{sync_id}"
+        while True:
```
Comment (AchuthaSourabhC) [inline dagster_rudderstack/resources/rudderstack.py:168]: would there be some timeout for the polling or would this wait as long as task runs ?

Comment (RanjeetMishra) [reply dagster_rudderstack/resources/rudderstack.py:168]: We can add poll timeout as optional.

Comment (RanjeetMishra) [reply dagster_rudderstack/resources/rudderstack.py:170]: Added polltimeout option.

File: setup.cfg (hunk 5 above)
Comment (RanjeetMishra) [reply setup.cfg]: Removed setup files. Using static version as version upgrades should not be frequent.

**Review by vyeshwanth — APPROVED** [review]

File: dagster_rudderstack/resources/rudderstack.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+                    timeout=self.request_timeout,
+                )
+                if data:
+                    request_args["json"] = data
+                response = requests.request(**request_args)
+                response.raise_for_status()
+                return response.json()
+            except requests.RequestException as e:
+                self._log.error(f"Request to url: {url} failed: {e}")
+                if num_retries == self.request_max_retries:
```
Comment (AchuthaSourabhC) [inline dagster_rudderstack/resources/rudderstack.py:109]: is there option for this to be infinite retry if customer doesn't want any ? could be default behavior

**Review by AchuthaSourabhC — APPROVED** [review]

File: dagster_rudderstack/resources/rudderstack.py (hunk 11 above)
Comment (RanjeetMishra) [reply dagster_rudderstack/resources/rudderstack.py:109]: `is there option for this to be infinite retry if customer doesn't want any` - No
Not sure if this is right to keep retries on error infinite. Better to let it fail after few retries and we can get issue from customer.

**Pull request merged**